### PR TITLE
receivers: populate author_count

### PR DESCRIPTION
**Pull request by chris-asl**
* Add author_count field in hep mapping.
* Populate author_count when `inspire_roles` doesn't contain
  `supervisor`.

Signed-off-by: Chris Aslanoglou <chris.aslanoglou@gmail.com>

## Motivation and Context
For adding support for `author-count` related queries.

I've also run the `prettify_json` script for the `holdingpen/hep` mapping. If we don't want it on this PR then I can create another one.

## Checklist:

- [x] I have all the information that I need (if not, move to `RFC` and look for it).
- [x] I linked the related issue(s) in the corresponding commit logs.
- [x] I wrote [good commit log messages](https://github.com/torvalds/subsurface-for-dirk/blob/5f15ad5a86ada3c5e574041a5f9d85235322dabb/README#L92-L119).
- [x] My code follows the code style of this project.
- [x] I've added any new docs if API/utils methods were added.
- [x] I have updated the existing documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: inspirehep/modules/workflows/mappings/holdingpen/hep.json [hunk 1]
```diff
@@ -23,11 +28,6 @@
                                 }
                             },
                             "type": "object"
-                        },
-                        "record_matches": {
-                            "include_in_all": false,
-                            "enabled": false,
-                            "type": "object"
```
Comment (jacquerie) [inline inspirehep/modules/workflows/mappings/holdingpen/hep.json]: This missing `prettify_json` should go in its own commit.

File: inspirehep/modules/records/receivers.py [hunk 2]
```diff
@@ -388,3 +388,14 @@ def generate_name_variations(sender, json, *args, **kwargs):
                 'output': full_name,
                 'payload': {'bai': bais[0] if bais else None}
             }})
+
+
+def populate_author_count(sender, json, *args, **kwargs):
```
Comment (jacquerie) [inline inspirehep/modules/records/receivers.py]: This needs to be added above to `enhance_after_index`: there's no magic here that collects the functions in this module and runs them, it's all very explicit...

File: inspirehep/modules/records/receivers.py [hunk 3]
```diff
@@ -388,3 +388,14 @@ def generate_name_variations(sender, json, *args, **kwargs):
...
+    """Populate author-count using only authors with 'inspi
+    re_roles' not equal to 'supervisor'."""
```
Comment (jacquerie) [inline inspirehep/modules/records/receivers.py]: Relevant portion of PEP 0257 (https://www.python.org/dev/peps/pep-0257/#one-line-docstrings):
> One-liners are for really obvious cases. They should really fit on one line.

File: inspirehep/modules/records/receivers.py [hunk 4]
```diff
@@ -388,3 +388,14 @@ def generate_name_variations(sender, json, *args, **kwargs):
...
+    if 'hep.json' not in json.get('$schema'):
+        return
+
+    authors = json.get('authors', [])
+
+    json['author_count'] = len([author for author in authors if 'supervisor' not in author.get('inspire_roles', [])])
```
Comment (jacquerie) [inline inspirehep/modules/records/receivers.py]: Let's split this expression over multiple lines with explaining variables (http://wiki.c2.com/?IntroduceExplainingVariable).

File: inspirehep/modules/records/receivers.py (hunk 2 above)
Comment (chris-asl) [reply inspirehep/modules/records/receivers.py]: Oops, forgot about that. Will amend.

File: tests/unit/records/test_records_receivers.py [hunk 5]
```diff
@@ -1145,3 +1146,33 @@ def test_populate_affiliation_suggest_does_nothing_if_record_is_not_institution(
     populate_affiliation_suggest(None, record)
 
     assert 'affiliation_suggest' not in record
+
+
+def test_populate_author_count():
+    schema = load_schema('hep')
+    schema_schema = schema['properties']['$schema']
```
Comment (jacquerie) [inline tests/unit/records/test_records_receivers.py]: This thing I did was a terrible idea... we can avoid adding assertions for this.

Comment (jacquerie) [reply tests/unit/records/test_records_receivers.py]: On the other hand, adding a test for the different `$schema` would be great.

File: inspirehep/modules/records/receivers.py (hunk 4 above)
Comment (chris-asl) [reply inspirehep/modules/records/receivers.py]: I agree! Does this look better 
```
authors_excluding_supervisors = [
    author for author in authors
    if 'supervisor' not in author.get('inspire_roles', [])
]
json['author_count'] = len(authors_excluding_supervisors)
```
?

**Comment by chris-asl** [comment]
@jacquerie I've integrated your changes. 
Also, I've added a last commit that fixes a "typo" in one of the `populate_earliest_date` related test.
